src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5570,36 +5511,40 @@ private static Task InternalWhenAll(Task[] tasks)
         // which involves several allocations, with this logic:
         //      return new WhenAllPromise(tasksCopy);
         // which saves a couple of allocations and enables debugger notification specialization.
-        //
-        // Used in InternalWhenAll(Task[])
         private sealed class WhenAllPromise : Task, ITaskCompletionAction
         {
-            /// <summary>
-            /// Stores all of the constituent tasks.  Tasks clear themselves out of this
-            /// array as they complete, but only if they don't have their wait notification bit set.
-            /// </summary>
-            private readonly Task?[] m_tasks;
             /// <summary>The number of tasks remaining to complete.</summary>
             private int m_count;
+            /// <summary>True if any of the supplied tasks require wait notification.</summary>
+            private bool m_anyTaskRequiresNotifyDebuggerOfWaitCompletion;
+            /// <summary>The tasks that faulted.</summary>
+            private ConcurrentQueue<Task>? m_faultedTasks;
+            /// <summary>The first task that got canceled.</summary>
+            private Task? m_canceledTask;
 
-            internal WhenAllPromise(Task[] tasks)
+            internal WhenAllPromise(IEnumerable<Task?>? tasks)
             {
-                Debug.Assert(tasks != null, "Expected a non-null task array");
-                Debug.Assert(tasks.Length > 0, "Expected a non-zero length task array");
+                if (tasks == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.tasks);
 
                 if (TplEventSource.Log.IsEnabled())
                     TplEventSource.Log.TraceOperationBegin(this.Id, "Task.WhenAll", 0);
 
                 if (s_asyncDebuggingEnabled)
                     AddToActiveTasks(this);
 
-                m_tasks = tasks;
-                m_count = tasks.Length;
-
-                foreach (Task task in tasks)
+                int count = 0;
+                foreach (Task? task in tasks)
                 {
+                    if (task == null) ThrowHelper.ThrowArgumentException(ExceptionResource.Task_MultiTaskContinuation_NullTask, ExceptionArgument.tasks);
                     if (task.IsCompleted) this.Invoke(task); // short-circuit the completion action, if possible
                     else task.AddCompletionAction(this); // simple completion action
+                    m_anyTaskRequiresNotifyDebuggerOfWaitCompletion |= task.IsWaitNotificationEnabled && task.ShouldNotifyDebuggerOfWaitCompletion; // potential recursion
+                    count += 1;
+                }
+
+                if ((m_count = Interlocked.Add(ref m_count, count)) == 0) // Condition is true if all tasks already completed
```

Comment:
Why is m_count being assigned here?  Presumably that was unintentional.

Reply:
Fixed.